pallets/relay-storage-roots/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//! One storage root is inserted per parachain block, but there may be more than one relay block in
+//! between two parachain blocks. In that case, there will be a gap in the `RelayStorageRoot`
+//! mapping. When creating a proof, users should ensure that they are using the latest storage root
+//! available in the mapping, otherwise it may not be possible to validate their proof.
+//!
+//! The `RelayStorageRoot` mapping is bounded by `MaxStorageRoots`. To ensure that oldest storage
+//! roots are removed first, there is an additional `RelayStorageRootKeys` storage item that stores
+//! a sorted list of all the keys. This is needed because it is not possible to iterate over a
+//! mapping in order (unless if using an `Identity` hash). The `MaxStorageRoots` limit applies to
+//! the number of items, not to their age.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::weights::WeightInfo;
+use cumulus_pallet_parachain_system::RelaychainStateProvider;
+use cumulus_primitives_core::relay_chain::BlockNumber as RelayBlockNumber;
+use frame_support::pallet;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_core::Get;
+use sp_core::H256;
+use sp_std::collections::vec_deque::VecDeque;
+
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod benchmarks;
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RelaychainStateProvider: RelaychainStateProvider;
+		#[pallet::constant]
+		type MaxStorageRoots: Get<u32>;
+		/// Weight info
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Map of relay block number to relay storage root
+	#[pallet::storage]
+	pub type RelayStorageRoot<T: Config> =
+		StorageMap<_, Twox64Concat, RelayBlockNumber, H256, OptionQuery>;
+
+	/// List of all the keys in `RelayStorageRoot`.
+	/// Used to remove the oldest key without having to iterate over all of them.
+	#[pallet::storage]
+	pub type RelayStorageRootKeys<T: Config> =
+		StorageValue<_, BoundedVec<RelayBlockNumber, T::MaxStorageRoots>, ValueQuery>;
+
+	impl<T: Config> Pallet<T> {
+		/// Populates `RelayStorageRoot` using `RelaychainStateProvider`.
+		pub fn set_relay_storage_root() {
+			let relay_state = T::RelaychainStateProvider::current_relay_chain_state();
+
+			// If this relay block number has already been stored, skip it.
+			if <RelayStorageRoot<T>>::contains_key(relay_state.number) {
+				return;
+			}
+
+			<RelayStorageRoot<T>>::insert(relay_state.number, relay_state.state_root);
+			let mut keys: VecDeque<_> = <RelayStorageRootKeys<T>>::get().into_inner().into();
+			keys.push_back(relay_state.number);
+			// Delete the oldest stored root if the total number is greater than MaxStorageRoots
+			if u32::try_from(keys.len()).unwrap() > T::MaxStorageRoots::get() {
```

Comment:
Please not use unwrap inside runtime production code, use an explicit saturation to u32::MAX instead

Reply:
Agree, thanks for spotting!